monitoring-dashboard/components/org.wso2.ei.dashboard.core/src/main/java/org/wso2/ei/dashboard/core/rest/delegates/heartbeat/HeartBeatDelegate.java

```diff
@@ -48,30 +47,34 @@ public class HeartBeatDelegate {
     private final DataManager dataManager = DataManagerSingleton.getDataManager();
 
     public Ack processHeartbeat(HeartbeatRequest heartbeatRequest) throws ManagementApiException {
-        long currentTimestamp = System.currentTimeMillis();
-        Ack ack = new Ack(Constants.FAIL_STATUS);
-        HeartbeatObject heartbeat = new HeartbeatObject(
-                heartbeatRequest.getProduct(), heartbeatRequest.getGroupId(), heartbeatRequest.getNodeId(),
-                heartbeatRequest.getInterval(), heartbeatRequest.getMgtApiUrl(), currentTimestamp);
-        if (logger.isDebugEnabled()) {
-            logger.debug("Management API URL received is: " + heartbeat.getMgtApiUrl());
-        }
-        boolean isSuccess;
-        String productName = heartbeat.getProduct();
-        ArtifactsManager artifactsManager = getArtifactManager(productName, heartbeat);
+        try {
+            long currentTimestamp = System.currentTimeMillis();
+            Ack ack = new Ack(Constants.FAIL_STATUS);
+            HeartbeatObject heartbeat = new HeartbeatObject(
+                    heartbeatRequest.getProduct(), heartbeatRequest.getGroupId(), heartbeatRequest.getNodeId(),
+                    heartbeatRequest.getInterval(), heartbeatRequest.getMgtApiUrl(), currentTimestamp);
+            if (logger.isDebugEnabled()) {
+                logger.debug("Management API URL received is: " + heartbeat.getMgtApiUrl());
+            }
+            boolean isSuccess;
+            String productName = heartbeat.getProduct();
+            ArtifactsManager artifactsManager = getArtifactManager(productName, heartbeat);
 
-        if (isNodeRegistered(heartbeat)) {
-            isSuccess = updateHeartbeat(heartbeat);
-        } else {
-            isSuccess = registerNode(heartbeat, artifactsManager);
+            if (isNodeRegistered(heartbeat)) {
+                isSuccess = updateHeartbeat(heartbeat);
+            } else {
+                isSuccess = registerNode(heartbeat, artifactsManager);
+                if (isSuccess) {
+                    artifactsManager.runFetchAllExecutorService();
+                }
+            }
             if (isSuccess) {
-                artifactsManager.runFetchAllExecutorService();
+                ack.setStatus(Constants.SUCCESS_STATUS);
             }
+            return ack;
+        } catch (DashboardServerException e) {
+            throw new ManagementApiException (e.getMessage(), 404);
```

Comment:
With this the exception stack trace will be lost?

Reply:
Added e.getCause() now.